core/pkg/filter/ast/lexer.go

```diff
@@ -258,25 +272,33 @@ func (s *scanner) scanToken() {
 			break
 		}
 
-		// TODO: We could return a more exact error message for Unicode chars if
-		// we added extra handling:
-		// https://stackoverflow.com/questions/53069040/checking-a-string-contains-only-ascii-characters
-		s.errors = append(s.errors, fmt.Errorf("unexpected character/byte at position %d. Please avoid Unicode.", s.nextByte-1))
+		// Check if the character is a Unicode character for a more precise error message
+		if c > 127 {
+			s.errors = append(s.errors, fmt.Errorf("unexpected Unicode character '%c' (U+%04X) at position %d", c, c, s.nextByte-1))
+		} else {
+			s.errors = append(s.errors, fmt.Errorf("unexpected character '%c' at position %d", c, s.nextByte-1))
+		}
 	}
 }
 
 // isIdentifierChar should match Kubernetes-supported name characters.
 //
 // https://kubernetes.io/docs/concepts/overview/working-with-objects/names/
 //
-// TODO: This may not match all characters we support for cluster IDs (it may be
-// the case that cluster IDs can contain UTF-8 characters).
-func isIdentifierChar(b byte) bool {
-	return (b >= '0' && b <= '9') || // 0-9
-		(b >= 'A' && b <= 'Z') || // A-Z
-		(b >= 'a' && b <= 'z') || // a-z
-		b == '-' || // hyphens are allowed according to K8s spec
-		b == '_' // underscores are allowed because of Prometheus sanitization
+// This has been updated to support UTF-8 characters for cluster IDs.
+func isIdentifierChar(r rune) bool {
+	// For ASCII characters, maintain the original logic
+	if r <= 127 {
+		b := byte(r)
+		return (b >= '0' && b <= '9') || // 0-9
+			(b >= 'A' && b <= 'Z') || // A-Z
+			(b >= 'a' && b <= 'z') || // a-z
+			b == '-' || // hyphens are allowed according to K8s spec
+			b == '_' // underscores are allowed because of Prometheus sanitization
+	}
+	
+	// For Unicode characters, allow letters and numbers
+	return unicode.IsLetter(r) || unicode.IsDigit(r)
```

Comment:
I don't think we need to check ascii specifically (the `unicode` checks should also check for ascii as well). We will have to alter the unicode checks to include punctuation as well. 

In general, we're a bit confused as to why we hard restricted the filter language fields and values to an ascii subset, so our thought is that we don't need any special ascii handling. What are your thoughts?

Reply:
Also worth noting that there are likely some tests that will fail (because the expectation was that we _only_ support a limited ascii character set). You can either repurpose those tests, or remove them if they're no longer required.